homeassistant/components/sun.py

```diff
@@ -51,172 +51,143 @@ def is_on(hass, entity_id=None):
     return hass.states.is_state(entity_id, STATE_ABOVE_HORIZON)
 
 
-def next_dawn(hass, entity_id=None):
+def next_dawn(hass, entity_id=None, offset=None):
     """Local datetime object of the next dawn.
 
     Async friendly.
     """
-    utc_next = next_dawn_utc(hass, entity_id)
+    utc_next = next_dawn_utc(hass, entity_id, offset)
 
     return dt_util.as_local(utc_next) if utc_next else None
 
 
-def next_dawn_utc(hass, entity_id=None):
+def next_dawn_utc(hass, entity_id=None, offset=None):
     """UTC datetime object of the next dawn.
 
     Async friendly.
     """
-    entity_id = entity_id or ENTITY_ID
-
-    state = hass.states.get(ENTITY_ID)
-
-    try:
-        return dt_util.parse_datetime(
-            state.attributes[STATE_ATTR_NEXT_DAWN])
-    except (AttributeError, KeyError):
-        # AttributeError if state is None
-        # KeyError if STATE_ATTR_NEXT_DAWN does not exist
+    entity = hass.data.get(entity_id or ENTITY_ID)
+    if entity is None:
         return None
 
+    return Sun.get_next_solar_event(
```

Comment:
Actually, I think @amelchio is on to something.

Automation: schedule 2 hours before sunset
 - If we start 1 hour before the next sunset, we schedule a "reschedule task" at sunset, because then the sun entity has been updated with the next sunset that we need.
 - When we are triggered at sunset we can then set the correct trigger.